Uses UTF8MB4 everywhere

Sesquipedalian opened the pull request:
Fixes #7938
Fixes #7173
Closes #6409
Closes #6406

- [Use InnoDB and dynamic rows for MySQL](https://github.com/SimpleMachines/SMF/commit/3a555440e1f9e4399c060cf924e5dbaad42b3ed3). This makes a whole lot of things easier. (See discussion in #7938.)
- [When installing, require utf8mb4 support for MySQL](https://github.com/SimpleMachines/SMF/commit/9cce0086c8f89c6610a04eca4d7ea8502f991885), for obvious reasons.
- [Create tables using utf8mb4 by default for MySQL](https://github.com/SimpleMachines/SMF/commit/d7c25ffaca88a8a7ccef101a7bfc1cff131a17fa) and [Always use utf8mb4 input/output for MySQL](https://github.com/SimpleMachines/SMF/commit/8a9edc8af7af24a29635ddd3cf051727bc4a2cf0), also for obvious reasons.
- [Autodetects $this->character_set and $this->mb4 in DB APIs](https://github.com/SimpleMachines/SMF/commit/2957e71a5d5e0d355dbd2f8d4d06fb8a71b0fae9). This means we no longer need to set those values in Settings.php.
- [Rewrites the ConvertUtf8 upgrader step for mb4 support](https://github.com/SimpleMachines/SMF/commit/4f21987e38297cbb757d41717ce1fea41b62b437). This is probably the most important part to test in multiple environments and with multiple datasets. It works for me, but I would ***really*** appreciate it if @sbulen could use that comprehensive test suite of his to test it.
- [Removes obsolete character set variables and related logic](https://github.com/SimpleMachines/SMF/commit/c5e5ce9e4333a9d9a41eda6e4de56b3d773cd0ff), because we don't need them anymore.
- [Moves Utils::fixUtf8Mb4() to DatabaseApi::fix_mb4()](https://github.com/SimpleMachines/SMF/commit/5578ae9cdfe2c851fcf3f42f2212e63aa8b7e43e) and [Transparently decodes entities for four-byte Unicode](https://github.com/SimpleMachines/SMF/pull/8425/commits/a5b106fe20d01296863b0dfd0138f21ef7db115f), because only the database API layer should know or care about whether the database supports 4-byte UTF-8. The rest of the application can and should work with the full UTF-8 character set without worrying about the details of the database's storage character set.

Sesquipedalian committed:
Moves Utils::fixUtf8Mb4() to DatabaseApi::fix_mb4()

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Autodetects $this->character_set and $this->mb4 in DB APIs

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Create tables using utf8mb4 by default for MySQL

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Removes obsolete character set variables and related logic

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Use InnoDB and dynamic rows for MySQL

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
When installing, require utf8mb4 support for MySQL

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Rewrites the ConvertUtf8 upgrader step for mb4 support

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Always use utf8mb4 input/output for MySQL

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

jdarwood007 commented:
Looks like 90% of this is just removing and hardcoding UTF-8 on everything.  InnoDB is a fairly safe conversion.  Just may take longer for larger forums on certain tables, but nothing can be avoided in timeout protections for that.   It looks good from what I see.

Sesquipedalian committed:
Transparently decodes entities for four-byte Unicode in Db\APIs\MySQL

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

sbulen commented:
I'll run some test upgrades.  Merge conflicts need to be resolved first, though.

Or is there a per-requisite PR?

Sesquipedalian committed:
Merge branch 'release-3.0' into 3.0/utf8mb4

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

# Conflicts:
#	Sources/Db/APIs/MySQL.php

Sesquipedalian commented:
> I'll run some test upgrades. Merge conflicts need to be resolved first, though.

Great! Thank you. 🙂 Merge conflict has been resolved.

> Or is there a per-requisite PR?

Nope.

sbulen commented:
First test was an upgrade of a new, vanilla 2.1.4 forum to 3.0, via CLI.  

I followed the old 2.1.x protocol, where I would copy the upgrade files over from the /other folder, then run upgrade.php.

DB: MySQL, version 8.4.0
php: 8.4.2

Had a few errors, here is the complete output:

>  * Updating Settings.php... Successful.
> **Error: Undefined variable $file_substrsteps File: D:\wamp64\www\van2130\upgrade.php Line: 2403** *  +++  Updating old values done.
>  +++  Changing default values done.
> . Successful.
>  *  +++  Removing all karma data, if selected done.
>  Successful.
>  *  +++  Emptying error log, if selected done.
>  Successful.
>  *  +++  Adding login history... done.
>  +++  Copying the current package backup setting... done.
>  +++  Copying the current "allow users to disable word censor" setting...**Error: Trying to access array offset on null File: D:\wamp64\www\van2130\upgrade.php(2521) : eval()'d code Line: 7** done.
>  +++  Converting collapsed categories... done.
>  +++  Parsing board descriptions and names done.
>  +++  Dropping "collapsed_categories" done.
>  +++  Adding new "topic_move_any" setting done.
>  +++  Adding new "enable_ajax_alerts" setting done.
>  +++  Adding new "alerts_auto_purge" setting done.
>  +++  Adding new "minimize_files" setting done.
>  +++  Collapse object done.
>  +++  Adding new "DEFAULTMaxListItems" setting done.
>  +++  Adding new "loginHistoryDays" setting done.
>  +++  Enable some settings we ripped from Theme settings done.
>  +++  Adding new "httponlyCookies" setting done.
>  +++  Adding new "samesiteCookies" setting done.
>  +++  Calculate appropriate hash cos**tError: Passing E_USER_ERROR to trigger_error() is deprecated since 8.4, throw an exception or call exit with a string message instead File: D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php Line: 2527Error: Invalid data structure sent to the database.<em>(upgrade.php(2521) : eval()'d code-1)</em> File: D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php Line: 2527**
> **Fatal error: Uncaught TypeError: array_combine(): Argument 2 ($values) must be of type array, string given in D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php on line 430**
> 
> **TypeError: array_combine(): Argument 2 ($values) must be of type array, string given in D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php on line 430**
> 
> Call Stack:
>     0.0079    1402080   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
>     0.4792    5966336   2. DatabaseChanges() D:\wamp64\www\van2130\upgrade.php:371
>     0.4833    5967968   3. parse_sql($filename = 'D:\\wamp64\\www\\van2130/upgrade_2-1_MySQL.sql') D:\wamp64\www\van2130\upgrade.php:1869
>     2.3360    6283880   4. eval('D:\wamp64\www\van2130\upgrade.php(2521) : eval()'d code') D:\wamp64\www\van2130\upgrade.php:2521
>     3.0980    6781696   5. SMF\Db\APIs\MySQL->insert($method = 'replace', $table = '{db_prefix}settings', $columns = ['variable' => 'string', 'value' => 'string'], $data = [0 => 'bcrypt_hash_cost', 1 => 14], $keys = [0 => 'variable'], $returnmode = ???, $connection = ???) D:\wamp64\www\van2130\upgrade.php(2521) : eval()'d code:1
>     3.0982    6782016   6. array_combine($keys = [0 => 'variable', 1 => 'value'], $values = 'bcrypt_hash_cost') D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php:430
>

sbulen commented:
I repeated the above test via browser, & get this error.  Note the 2.1 environment being upgraded is using the default, English.  

This may help understand the CLI errors:

![pr8425_upgr_msg](https://github.com/user-attachments/assets/8c09c875-3af2-4996-a995-6daec48be968)

Sesquipedalian commented:
Okay, so it looks like we have some unrelated upgrader bugs to fix before you can even get to the point of testing the new ConvertUtf8() logic in this PR.

Oh, the joys of the upgrader never cease. 😒

sbulen commented:
Looks like 2 things here...
 - 8.4 (https://github.com/SimpleMachines/SMF/issues/8402)
 - Different language identifiers from 2.1 to 3...

sbulen commented:
Probably the same issues in a different form... But I attempted an install in the same environment.

This comes up after entering the DB credentials, etc.   Same issues in php 8.3 & 8.4:
> Warning: Undefined array key 0 in D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php on line 915
> 
> Warning: Trying to access array offset on null in D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php on line 915
> 
> Fatal error: Uncaught TypeError: SMF\Db\APIs\MySQL::detect_charset(): Return value must be of type string, null returned in D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php:934 Stack trace: 0 D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php(2195): SMF\Db\APIs\MySQL->detect_charset('messages', 'body') 1 D:\wamp64\www\van2130\Sources\Db\DatabaseApi.php(331): SMF\Db\APIs\MySQL->__construct(Array) # D:\wamp64\www\van2130\install.php(904): SMF\Db\DatabaseApi::load(Array) 3 D:\wamp64\www\van2130\install.php(172): DatabaseSettings() 4 {main} thrown in D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php on line 934

sbulen commented:
Same errors occur in unix.

Sesquipedalian committed:
Handles non-existent databases in Db::detect_charset()

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Merge branch 'release-3.0' into 3.0/utf8mb4

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

# Conflicts:
#	Sources/Actions/AttachmentDownload.php
#	Sources/Subs-Compat.php
#	Sources/Utils.php
#	Themes/default/Display.template.php
#	other/upgrade.php

Sesquipedalian commented:
> Warning: Undefined array key 0 in D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php on line 915
> 
> Warning: Trying to access array offset on null in D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php on line 915
> 
> Fatal error: Uncaught TypeError: SMF\Db\APIs\MySQL::detect_charset(): Return value must be of type string, null returned in D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php:934 Stack trace: 0 D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php(2195): SMF\Db\APIs\MySQL->detect_charset('messages', 'body') 1 D:\wamp64\www\van2130\Sources\Db\DatabaseApi.php(331): SMF\Db\APIs\MySQL->__construct(Array) # D:\wamp64\www\van2130\install.php(904): SMF\Db\DatabaseApi::load(Array) 3 D:\wamp64\www\van2130\install.php(172): DatabaseSettings() 4 {main} thrown in D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php on line 934

I've added a fix for this one in 28b8f46. 

The other errors appear to be unrelated to this PR, so I will submit a new PR with fixes for them. We'll need to deal those first, and the come back to this PR after.

Sesquipedalian commented:
PR with fixes for the unrelated errors: #8437

sbulen commented:
A heads up - there are now some merge errors here that need to be addressed before I can resume testing.

Sesquipedalian committed:
Merge branch 'release-3.0' into 3.0/utf8mb4

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

# Conflicts:
#	other/upgrade.php

Sesquipedalian commented:
Should be ready now, @sbulen. 🙂

sbulen commented:
**Just attempted an initial install - not getting very far:**

> Warning: Undefined array key 0 in D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php on line 921
> 
> Warning: Trying to access array offset on null in D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php on line 921
> 
> Fatal error: Uncaught TypeError: SMF\Db\APIs\MySQL::detect_charset(): Return value must be of type string, null returned in D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php:940 Stack trace: 0 D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php(2201): SMF\Db\APIs\MySQL->detect_charset('messages', 'body') 1 D:\wamp64\www\van2130\Sources\Db\DatabaseApi.php(331): SMF\Db\APIs\MySQL->__construct(Array) 2 D:\wamp64\www\van2130\install.php(904): SMF\Db\DatabaseApi::load(Array) 3 D:\wamp64\www\van2130\install.php(172): DatabaseSettings() 4 {main} thrown in D:\wamp64\www\van2130\Sources\Db\APIs\MySQL.php on line 940

**Also attempted a 2.1 => 3.0 upgrade, which got pretty far, but had these errors.  Partial log:**

 _*  +++  Adding a new column "spoofdetector_name" to members table done.
 +++  Adding new "spoofdetector_censor" setting done.
 Successful.
 *  +++  Adding a new column "smf_version" to log_packages table done.
 Successful.
.
Warning: Undefined variable $table_info in D:\wamp64\www\van2130\upgrade.php on line 3761

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990


Warning: Trying to access array offset on null in D:\wamp64\www\van2130\upgrade.php on line 3761

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990

Converting table  to utf8mb4... done.

Warning: Undefined variable $table_info in D:\wamp64\www\van2130\upgrade.php on line 3761

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990


Warning: Trying to access array offset on null in D:\wamp64\www\van2130\upgrade.php on line 3761

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990

Converting table  to utf8mb4... done.

Warning: Undefined variable $table_info in D:\wamp64\www\van2130\upgrade.php on line 3761

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990


Warning: Trying to access array offset on null in D:\wamp64\www\van2130\upgrade.php on line 3761

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990

Converting table  to utf8mb4... done.

Warning: Undefined variable $table_info in D:\wamp64\www\van2130\upgrade.php on line 3761

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990


Warning: Trying to access array offset on null in D:\wamp64\www\van2130\upgrade.php on line 3761

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990

Converting table  to utf8mb4... done.
.
Warning: Undefined variable $table_info in D:\wamp64\www\van2130\upgrade.php on line 3761

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990


Warning: Trying to access array offset on null in D:\wamp64\www\van2130\upgrade.php on line 3761

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990

Converting table  to utf8mb4... done.

Warning: Undefined variable $table_info in D:\wamp64\www\van2130\upgrade.php on line 3761

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990


Warning: Trying to access array offset on null in D:\wamp64\www\van2130\upgrade.php on line 3761

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990

Converting table  to utf8mb4... done.

Warning: Undefined variable $table_info in D:\wamp64\www\van2130\upgrade.php on line 3761

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990


Warning: Trying to access array offset on null in D:\wamp64\www\van2130\upgrade.php on line 3761

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990

Converting table  to utf8mb4... done.

Warning: Undefined variable $table_info in D:\wamp64\www\van2130\upgrade.php on line 3761

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990


Warning: Trying to access array offset on null in D:\wamp64\www\van2130\upgrade.php on line 3761

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990

Converting table  to utf8mb4...
Warning: Undefined array key "charset" in D:\wamp64\www\van2130\upgrade.php on line 3767

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990


Warning: Undefined array key "charset" in D:\wamp64\www\van2130\upgrade.php on line 3791

Call Stack:
    0.0079    1402368   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
   14.1515    8791872   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
   14.1515    8791872   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3990

Cannot convert string '\x80' from binary to utf8mb4_

Sesquipedalian committed:
Handles non-existent databases in Db::detect_charset(), for real

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Fixes an incorrect variable name in upgrader's ConvertUtf8()

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Don't try to convert charsets of columns that don't have them

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian commented:
Ready for another run, @sbulen.

sbulen commented:
**Initial install results - it looks like the install completes, but pulling up the app, I get this:**

> Fatal error: Uncaught TypeError: SMF\Utils::entityDecode(): Argument 2 ($flags) must be of type int, true given, called in D:\wamp64\www\van2130\Sources\Slug.php on line 275 and defined in D:\wamp64\www\van2130\Sources\Utils.php:358 Stack trace: #0 D:\wamp64\www\van2130\Sources\Slug.php(275): SMF\Utils::entityDecode('Welcome to SMF!', true) 1 D:\wamp64\www\van2130\Sources\Slug.php(185): SMF\Slug->build('Welcome to SMF!', 30) 2 D:\wamp64\www\van2130\Sources\Slug.php(227): SMF\Slug->__construct('Welcome to SMF!', 'topic', 1, 30) 3 D:\wamp64\www\van2130\Sources\Actions\BoardIndex.php(377): SMF\Slug::create('Welcome to SMF!', 'topic', 1) 4 D:\wamp64\www\van2130\Sources\Actions\BoardIndex.php(67): SMF\Actions\BoardIndex::get(Array) 5 D:\wamp64\www\van2130\Sources\Forum.php(489): SMF\Actions\BoardIndex->execute() 6 D:\wamp64\www\van2130\index.php(149): SMF\Forum->execute() 7 {main} thrown in D:\wamp64\www\van2130\Sources\Utils.php on line 358

**Note, I'm not sure if this is intended or not, but the default collation is utf8mb4_0900_ai_ci:**
![image](https://github.com/user-attachments/assets/0072ea96-6c25-4111-94d7-ce935a8f9720)

**I also attempted an upgrade, but it failed during the utf8mb4 conversion.  Partial log:**

> Converting table smf_ban_groups to utf8mb4... done.
> .Converting table smf_ban_items to utf8mb4... done.
> Converting table smf_board_permissions to utf8mb4... done.
> Converting table smf_boards to utf8mb4... done.
> Converting table smf_calendar to utf8mb4...You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '=' at line 2

Sesquipedalian committed:
Utils::entityDecode() doesn't need this param anymore

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian commented:
> Fatal error: Uncaught TypeError: SMF\Utils::entityDecode(): Argument 2 ($flags) must be of type int, true given, called in D:\wamp64\www\van2130\Sources\Slug.php on line 275 and defined in D:\wamp64\www\van2130\Sources\Utils.php:358

Fixed now. SMF\Slug was written in parallel to this PR, so it didn't reflect the removal of the $mb4 parameter from Utils::entityDecode().

> Note, I'm not sure if this is intended or not, but the default collation is utf8mb4_0900_ai_ci:

That is intentional. utf8mb4_0900_ai_ci is the recommended and best general collation for UTF-8, because it based on more recent Unicode collation rules than any other option. See https://dev.mysql.com/doc/refman/8.4/en/charset-collation-names.html for more info.

> Converting table smf_calendar to utf8mb4...You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '=' at line 2

Hm. That's not a very helpful message, unfortunately. Line 2 of what? I'll try to track it down.

Sesquipedalian commented:
Here, @sbulen, please try with this modified copy of upgrade.php. It is identical to the version in 7ddc08c, except that it will dump info about each column during the attempt to convert it to utf8mb4. That info will help me figure out which column is causing the error.

[upgrade.txt](https://github.com/user-attachments/files/18697344/upgrade.txt)

sbulen commented:
Oh the joys of testing the upgrader!  Seriously, at the very least, it needs a log like you can get from CLI.

Output:

> Converting table smf_board_permissions to utf8mb4... done.
> Converting table smf_boards to utf8mb4... done.
> Converting table smf_calendar to utf8mb4...1: [
> 	'name' => 'title',
> 	'not_null' => true,
> 	'null' => false,
> 	'default' => '',
> 	'type' => 'varchar',
> 	'size' => '255',
> 	'auto' => false,
> 	'charset' => 'utf8mb3',
> ]
> You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '=' at line 2

I think a clue may be found in the very first row in the table:
![image](https://github.com/user-attachments/assets/5022bf4d-ad6c-4d5c-9b1d-ffec52222da3)

Sesquipedalian committed:
Adds missing quotes around default values in ConvertUtf8()

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian commented:
> You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '=' at line 2

It appears to have been just a thickheaded mistake on my part. 🙄

Sesquipedalian committed:
Don't skip character 0xFF in source charset when converting to UTF-8

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

sbulen commented:
These things happen.  It's a lot of work.  Let me know when I should retest.

My availability the next couple of days is very spotty, and I'm still fixing/restoring the set of test environments.   I'm also building new environments that include things like older row_formats.  So it may slow down a bit for a couple days.

Sesquipedalian commented:
Go ahead as soon as you can. Thanks again, @sbulen.

sbulen commented:
The latest:
3.0 Installer OK 🍻
2.1 to 3.0 upgrade 😥

Partial log:

> Converting table smf_ban_groups to utf8mb4... done.
> Converting table smf_ban_items to utf8mb4... done.
> .Converting table smf_board_permissions to utf8mb4... done.
> Converting table smf_boards to utf8mb4... done.
> Converting table smf_calendar to utf8mb4...You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '=''' at line 2

sbulen commented:
OK another test, interesting results...

Source DB is a 2.0 forum, utf8mb3, InnoDB, row_format = COMPACT.   I wanted to ensure the row_format got converted.  I believe this would match a forum that converted themselves to InnoDB in the early InnoDB days, ~MySQL 5.5.

This fails in the 2.1 step...
>  +++  Adjusting the indexes for log_group_requests done.
>  Successful.
>  *  +++  Adding support for <credits> tag in package manager done.
>  +++  Adding support for package hashesUnsuccessful!  Database error message:
> Row size too large. The maximum row size for the used table type, not counting BLOBs, is 8126. This includes storage overhead, check the manual. You have to change some columns to TEXT or BLOBs

I have not seen this error in the support forum.  Which tells me that folks who are still using row_format != dynamic are kinda rare, &/or, they figure out themselves how to get past this.

I wonder if we need to do the engine/row_format change up front?

Prerequisites are prerequisites...

Sesquipedalian committed:
Uses Db::change_column() in ConvertUtf8()

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian commented:
> Converting table smf_calendar to utf8mb4...You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '=''' at line 2

It occurred to me that we could just use Db::change_column() to do this, rather than the custom query I was using before. That will likely avoid the tedious debugging I would otherwise need to do.

Sesquipedalian committed:
Merge branch 'release-3.0' into 3.0/utf8mb4

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Convert to InnoDB before making other database changes

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian commented:
> I wonder if we need to do the engine/row_format change up front?

Good idea. Hopefully [5d2966a](https://github.com/SimpleMachines/SMF/pull/8425/commits/5d2966adaf842d78f79dc2b86f866798ccd41413) will fix this issue for us.

sbulen commented:
FYI, there are now merge conflicts with this PR.

Sesquipedalian committed:
Merge branch 'release-3.0' into 3.0/utf8mb4

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian commented:
Conflicts have been resolved. This is ready for another round of testing whenever you are, @sbulen. 🙂

Sesquipedalian committed:
Temp fix for a weird error in the upgrader

I don't feel inclined to hunt down the root cause of this since we will soon be replacing the entire upgrader anyway.

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian commented:
@sbulen, I've added a temporary fix that should allow you to get past the issue that you ran into with 3.0 → 3.0 upgrades and mentioned in https://github.com/SimpleMachines/SMF/pull/8459#issuecomment-2652508481. It's just a bandaid, since the underlying problem raises questions about why `$upcontext['language']` wasn't set already at that point, but it's not worth hunting that down now when we are going to replace the entire upgrader soon anyway.

Sesquipedalian committed:
Merge branch 'release-3.0' into 3.0/utf8mb4
Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Only redirect to HTTPS if $_SERVER['REQUEST_URL'] starts with http://

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

sbulen commented:
3.0 Install 🍻 OK
2.1 => 3.0 - a few issues....

The upgrader wouldn't run in CLI mode.  No error, no nothing, immediate exit.

In browser mode, I get an error:
> The upgrader could not find the "Install" language file for your selected language, . Upgrade will continue with the forum default, en_US. 

If I ignore & attempt to proceed, it immediately WSOD's with:
> Unknown column 'uid' in 'field list'

Sesquipedalian committed:
Don't reconvert to utf8mb4 unnecessarily

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian commented:
> The upgrader wouldn't run in CLI mode. No error, no nothing, immediate exit.

[98dab02](https://github.com/SimpleMachines/SMF/pull/8425/commits/98dab02838273310ed3510d4cb6e62d65979d092) fixes that.

> In browser mode, I get an error:
> 
> > The upgrader could not find the "Install" language file for your selected language, . Upgrade will continue with the forum default, en_US.
> 
> If I ignore & attempt to proceed, it immediately WSOD's with:
> 
> > Unknown column 'uid' in 'field list'

Hm. I'm not sure about that one yet. I'll look into it next.

Sesquipedalian committed:
Don't get stuck waiting for conversion to utf8mb4 to complete

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Fixes user language lookup in upgrader

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Populates calendar table's uid column in the SQL files

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Don't get stuck waiting for ConvertToUtf8(), for real this time

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Fixes error when there are security warnings but not about files

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian commented:
Okay, @sbulen, ready for another round. All upgrader errors introduced since merging from release-3.0 have been fixed according my tests. Here's hoping that it passes your tests, too.

sbulen commented:
MySQL 8.4, PHP 8.4.2.

3.0 install was fine.  

Just attempted a 3.0 => 3.0 upgrade & got the same as before - CLI produces an mmediate exit; when run from the browser you get a "try again" link that doesn't work.

Also attempted a 2.1 => 3.0 upgrade & got this (partial log; 2.1 was in utf8mb3):
> .Converting table smf_boards to utf8mb4... done.
> Converting table smf_calendar to utf8mb4...
> Fatal error: Uncaught ValueError: mb_convert_encoding(): Argument 3 ($from_encoding) contains invalid encoding "utf8mb3" in D:\wamp64\www\van2130\upgrade.php on line 3793
> 
> ValueError: mb_convert_encoding(): Argument  3 ($from_encoding) contains invalid encoding "utf8mb3" in D:\wamp64\www\van2130\upgrade.php on line 3793
> 
> Call Stack:
>     0.0078    1397320   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
>    17.5674    8419496   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
>    17.5674    8419496   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3968
>    20.5053    8440984   4. mb_convert_encoding($string = '\000', $to_encoding = 'UTF-8', $from_encoding = 'utf8mb3') D:\wamp64\www\van2130\upgrade.php:3793
> 
>

Sesquipedalian commented:
> Just attempted a 3.0 => 3.0 upgrade & got the same as before - CLI produces an mmediate exit; when run from the browser you get a "try again" link that doesn't work.

I cannot reproduce that. I've tried both the CLI and in browser repeatedly on an existing 3.0 database.

What output do you see if you make the following temporary change in QueryString and then try again using CLI?

Find:
```
	/**
	 * Checks to see if we're forcing SSL, and redirects if necessary.
	 */
	protected static function sslRedirect(): void
	{
```

Replace:
```
	/**
	 * Checks to see if we're forcing SSL, and redirects if necessary.
	 */
	protected static function sslRedirect(): void
	{
		echo $_SERVER['REQUEST_URL'] ?? 'derp';
		die();
```

Sesquipedalian committed:
Changes 'utf8mb3' to 'UTF-8' in param for mb_convert_encoding()

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian commented:
> ValueError: mb_convert_encoding(): Argument 3 ($from_encoding) contains invalid encoding "utf8mb3" in D:\wamp64\www\van2130\upgrade.php on line 3793

Odd. I was also testing 2.1 → 3.0 using a database in utf8mb3, and I never saw that. Still, dc702ae should fix that.

Sesquipedalian committed:
Don't clobber the privacy policy when upgrading to 3.0

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Don't set spurious previousCharacterSet during 3.0 → 3.0 upgrades

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

sbulen commented:
Hmmm...  I'm still seeing this on a 2.1 => 3.0 upgrade:

> ...
> Converting table smf_calendar to utf8mb4...
> Fatal error: Uncaught ValueError: mb_convert_encoding(): Argument 3 ($from_encoding) contains invalid encoding "utf8mb4" in D:\wamp64\www\van2130\upgrade.php on line 3795
> 
> ValueError: mb_convert_encoding(): Argument 3 ($from_encoding) contains invalid encoding "utf8mb4" in D:\wamp64\www\van2130\upgrade.php on line 3795
> 
> Call Stack:
>     0.0078    1401488   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
>    17.6087    8423960   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
>    17.6087    8423960   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3970
>    22.3724    8445480   4. mb_convert_encoding($string = '\000', $to_encoding = 'UTF-8', $from_encoding = 'utf8mb4') D:\wamp64\www\van2130\upgrade.php:3795
> 
>

sbulen commented:
I believe MySQL is still in a kind of limbo...  For DB functions, I don't think we should ever use 'utf8', only 'utf8mb3' and 'utf8mb4'.  

My understanding is that the very definition of 'utf8' will change.  In 'older' versions of MySQL it means 'utf8mb3'.  At some point it will mean 'utf8mb4'.  I don't know when that cutoff will be.  But the point is that, _across mysql versions_, _**utf8 is an ambiguous term**_.

I am running MySQL 8.4.  I believe that in 8.4, 'utf8' as an alias for 'utf8mb3' is deprecated.
https://dev.mysql.com/doc/refman/8.4/en/charset-unicode-utf8.html

This kinda sorta explains that 'utf8' only appears in SHOW statements, etc:
https://www.percona.com/blog/the-power-of-utf8mb4-in-mysql-8-0-unleashing-the-full-potential-of-multilingual-data/

Sesquipedalian committed:
Changes all utf8* variants to UTF-8 in param for mb_convert_encoding()

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian commented:
> Hmmm... I'm still seeing this on a 2.1 => 3.0 upgrade:
> 
> > ...
> > Converting table smf_calendar to utf8mb4...
> > Fatal error: Uncaught ValueError: mb_convert_encoding(): Argument 3 ($from_encoding) contains invalid encoding "utf8mb4" in D:\wamp64\www\van2130\upgrade.php on line 3795
> > ValueError: mb_convert_encoding(): Argument 3 ($from_encoding) contains invalid encoding "utf8mb4" in D:\wamp64\www\van2130\upgrade.php on line 3795
> > Call Stack:
> > 0.0078    1401488   1. {main}() D:\wamp64\www\van2130\upgrade.php:0
> > 17.6087    8423960   2. serialize_to_json() D:\wamp64\www\van2130\upgrade.php:371
> > 17.6087    8423960   3. ConvertUtf8() D:\wamp64\www\van2130\upgrade.php:3970
> > 22.3724    8445480   4. mb_convert_encoding($string = '\000', $to_encoding = 'UTF-8', $from_encoding = 'utf8mb4') D:\wamp64\www\van2130\upgrade.php:3795

Please try b7db1fe.



> I believe MySQL is still in a kind of limbo... For DB functions, I don't think we should ever use 'utf8', only 'utf8mb3' and 'utf8mb4'.
> 
> My understanding is that the very definition of 'utf8' will change. In 'older' versions of MySQL it means 'utf8mb3'. At some point it will mean 'utf8mb4'. I don't know when that cutoff will be. But the point is that, _across mysql versions_, _**utf8 is an ambiguous term**_.
> 
> I am running MySQL 8.4. I believe that in 8.4, 'utf8' as an alias for 'utf8mb3' is deprecated. https://dev.mysql.com/doc/refman/8.4/en/charset-unicode-utf8.html
> 
> This kinda sorta explains that 'utf8' only appears in SHOW statements, etc: https://www.percona.com/blog/the-power-of-utf8mb4-in-mysql-8-0-unleashing-the-full-potential-of-multilingual-data/

The issue we are hitting right now (and that b7db1fe should solve properly) is not with the database but rather with mb_convert_encoding().  Most of the character set names that MySQL uses are recognized by mb_convert_encoding(), but not `utf8mb3` and `utf8mb4`.

Sesquipedalian commented:
Actually, wait. I have an even better idea than b7db1fe.

Sesquipedalian committed:
Don't perform byte-level conversion if already in UTF-8

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian commented:
Okay, [6952030](https://github.com/SimpleMachines/SMF/pull/8425/commits/69520306cbf975069465b5accea77bfb970c1ced) should be a better solution than [b7db1fe](https://github.com/SimpleMachines/SMF/pull/8425/commits/b7db1fe318e1422777e915d9d98c0ee7e3253e6e). The reason it is better is that if `$from_charset` is already some variant of UTF-8 then the byte-level conversion is entirely unnecessary.

Sesquipedalian commented:
Sorry if I'm doing that thing again where I change stuff while you are in the middle of testing. 🤪

sbulen commented:
No prob - I was doing some reading...

Just a couple notes on the old logic, since the code doesn't look very familiar to me anymore...
 - The old 2.1 logic did the check & the conversion column by column, in case things choked & the user had to restart.  Keep reruns/restarts in mind.
 - The old 2.1 logic, I believe, avoided using mb_convert_encoding because it is prone to double-encoding (do a quick search on mb_convert_encoding() & double encoding...).  Doing this 100% within MySQL helped, because MySQL is simply better at conversion than php.  That's why it's bounced off of binary, to force MySQL to detect the source charset - which it does very well.

The problem is some of the old MySQL versions were horrible at enforcing charsets.  So, there is a lot of utf8 data stuffed into old latin1 DBs.  Just running a simple utf8 conversion double-encoded everything.  The bounce-it-off-of-binary approach addresses that.

Yes, I'm always kicking the posts on Chesterton's fence...

Sesquipedalian commented:
I don't think I said this clearly above, but this is ready for another round of testing whenever you are, @sbulen.

Sesquipedalian commented:
> * The old 2.1 logic did the check & the conversion column by column, in case things choked & the user had to restart.  Keep reruns/restarts in mind.

I don't think that's an issue. Since database transactions are always atomic, changing the whole table at once whenever possible is actually better and safer. When we do one column at a time using the method where we change the column to a binary encoding and then back, an interruption at an inopportune moment can leave the column sitting there in a binary encoding. If anything, we should probably build more safety checks around the column-based method in the new upgrader code.



> * The old 2.1 logic, I believe, avoided using mb_convert_encoding because it is prone to double-encoding (do a quick search on mb_convert_encoding() & double encoding...).  Doing this 100% within MySQL helped, because MySQL is simply better at conversion than php.

I don't think that's accurate. The old logic was inherited from SMF 2.0, when the mbstring extension was not required by SMF and therefore the upgrader couldn't rely on it. 

Regarding double encoding, all that I found on the matter was a single StackOverflow discussion in which the original poster was trying to do something different than we are (and didn't seem to understand what they were doing very well). Perhaps your searches turned up something mine didn't, though, so if there's something more, please share the link. I am often wrong, after all, and always glad to discover a better understanding. 🙂

Sesquipedalian committed:
Be explicit about utf8mb3 in $charset_maps

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian commented:
> For DB functions, I don't think we should ever use 'utf8', only 'utf8mb3' and 'utf8mb4'.

I was just looking over the code again and noticed the spot that you were probably referring to; 14ea6ae should fix it.

Again, ready for testing whenever you are. 🙂

sbulen commented:
Back in town & will get back to testing.  Juggling a lot, so it may take a while.

A couple notes on the above...  First, this logic does go column by column when bouncing off of binary... See lines 3758+.

Also, mb_convert_encoding pretty much only does what you tell it to, and sometimes that's a bad thing...  The whole point of bouncing off of binary is to make use of MySQL's charset detection, which is far better than PHPs.   The real issue is a lot of legacy latin1, win1251, etc., with various other encodings stuffed into it, from back in the day when php & mysql were kinda awful at it...

Run this:
```
<?php
$text = "this is Middle English: He wes Leovenaðes sone -- liðe him be Drihten. \n";
echo "before... " . $text;
echo "after.... " . mb_convert_encoding($text, 'ISO-8859-1', 'UTF-8');
```

Food for thought.

Sesquipedalian committed:
Let MySQL handle charset detection internally when possible

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian commented:
I've made a change in order to let MySQL handle character set detection internally whenever and wherever possible. We now only do it manually for character sets that MySQL does not have native support for. I believe this will address your concerns, @sbulen.

Sesquipedalian committed:
Fixes inverted logic for filtering out unsupported charsets

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian commented:
@sbulen, have you had a chance to test the latest changes yet?

sbulen commented:
Latest round.  This time I'll include a list of all the tests I hope to power thru, including my technologically sophisticated result conveyance system:

3.0 Install	🍻 OK
3.0 => 3.0	FAIL 😥
2.1 => 3.0	🍻 OK
2.0 Latin1 => 3.0	FAIL 😥
2.0 Utf8 => 3.0	🍻 OK
2.0 Latin2 (Slovenian)	Not Yet Run 😐
1.1 => 3.0	Not Yet Run 😐
1.0 => 3.0	Not Yet Run 😐
Yabbse => 3.0	Not Yet Run 😐
Inst/Upgr DB Compare	Not Yet Run 😐
Confirm encoding	🍻 OK
Confirm InnoDB	🍻 OK
Confirm Dynamic	🍻 OK
pg 2.1 => 3.0	Not Yet Run 😐

**_The 3.0 => 3.0 result - it just stares at me & does not proceed:_**
![image](https://github.com/user-attachments/assets/6116ea50-f1e9-4198-befa-3f1f83015ec3)


**_The 2.0 latin1 => 3.0 result (Partial log... Note it's not unusual for display fields to fail to unserialize here - not sure why actually, it's been a while...  The real problem is the incorrect string value):_**
 *  +++  Adding a new column "spoofdetector_name" to members table done.
. +++  Adding new "spoofdetector_censor" setting done.
 Successful.
 *  +++  Adding a new column "smf_version" to log_packages table done.
 Successful.
 *  +++  Updating primary key for log_search_results table done.
 Successful.
 *  +++  Update mail_type Successful.
Converting data from serialize() to json_encode().
Fixing some settings...
 - Failed to unserialize the 'displayFields' setting. Skipping. done.
Successful.
Converting table rul_admin_info_files to utf8mb4...Incorrect string value: '\x967.2.'...' for column 'data' at row 3

sbulen commented:
Note the 2.0 file is just the standard admin info files, e.g.:

![image](https://github.com/user-attachments/assets/993a7cdb-6a6c-4ddd-9074-3e7ee84e8b57)

It's not liking the BLOBs...

sbulen commented:
OK, I snuck in a few more tests (yabb, 1.0 & 1.1), this is all I can do for today:

3.0 Install	OK 🍻
3.0 => 3.0	FAIL 😥
2.1 => 3.0	OK 🍻
2.0 Latin1 => 3.0	FAIL 😥
2.0 Utf8 => 3.0	OK 🍻
2.0 Latin2 (Slovenian)	Not Yet Run 😐
1.1 => 3.0	FAIL 😥
1.0 => 3.0	FAIL 😥
Yabbse => 3.0	FAIL 😥
Inst/Upgr DB Compare	Not Yet Run 😐
Confirm encoding	OK 🍻
Confirm InnoDB	OK 🍻
Confirm Dynamic	OK 🍻
pg 2.1 => 3.0	Not Yet Run 😐

_**1.1 issue:**_

> 
>  *  +++  Updating primary key for log_search_results table done.
> . Successful.
>  *  +++  Update mail_type Successful.
> Converting data from serialize() to json_encode().
> Fixing some settings...
>  - Failed to unserialize the 'displayFields' setting. Skipping. done.
> Successful.
> Converting table smf_attachments to utf8mb4... done.
> .Converting table smf_ban_groups to utf8mb4... done.
> Converting table smf_ban_items to utf8mb4... done.
> .Converting table smf_board_permissions to utf8mb4... done.
> Converting table smf_boards to utf8mb4... done.
> .Converting table smf_calendar to utf8mb4... done.
> .Converting table smf_categories to utf8mb4... done.
> .Converting table smf_log_actions to utf8mb4... done.
> .Converting table smf_log_banned to utf8mb4... done.
> Converting table smf_log_errors to utf8mb4...Invalid default value for 'session'

_**1.0 issue, very similar to 1.1:**_

>  *  +++  Update mail_type Successful.
> Converting data from serialize() to json_encode().
> Fixing some settings...
>  - Failed to unserialize the 'cal_today_birthday' setting. Skipping.
>  - Failed to unserialize the 'cal_today_event' setting. Skipping.
>  - Failed to unserialize the 'cal_today_holiday' setting. Skipping.
>  - Failed to unserialize the 'displayFields' setting. Skipping. done.
> Successful.
> Converting table smf_attachments to utf8mb4... done.
> .Converting table smf_board_permissions to utf8mb4... done.
> .Converting table smf_boards to utf8mb4... done.
> .Converting table smf_calendar to utf8mb4... done.
> .Converting table smf_categories to utf8mb4... done.
> Converting table smf_log_actions to utf8mb4... done.
> .Converting table smf_log_banned to utf8mb4... done.
> Converting table smf_log_errors to utf8mb4...Invalid default value for 'session'

_**Yabbse issue - didn't get very far at all...  (Yabbse is so different, I wonder if it's time to punt on this support...):**_

>  * Updating Settings.php... FAILURE.

sbulen commented:
Uh...  The 1.0 & 1.1 default for session is... uh... odd:

**_In DB:_**
![image](https://github.com/user-attachments/assets/275e08ce-ad2c-47b5-aa6c-96cba7affb16)

**_In Installer:_**
```
#
# Table structure for table `log_errors`
#

CREATE TABLE {$db_prefix}log_errors (
  ID_ERROR mediumint(8) unsigned NOT NULL auto_increment,
  logTime int(10) unsigned NOT NULL default '0',
  ID_MEMBER mediumint(8) unsigned NOT NULL default '0',
  ip char(16) NOT NULL default '                ',
  url text NOT NULL,
  message text NOT NULL,
  session char(32) NOT NULL default '                                ',
  PRIMARY KEY (ID_ERROR),
  KEY logTime (logTime),
  KEY ID_MEMBER (ID_MEMBER),
  KEY ip (ip(16))
) ENGINE=MyISAM;

```

sbulen commented:
All tests run in php 8.4.2 & mysql 8.4.0.

Oldiesmann commented:
We can look into the issue with YaBB SE, but definitely not something we need to spend a lot of time on. It's been 21 years since YaBB SE 1.5.5 was released and a little over 20 since SMF 1.0 was released (YaBB SE 1.5.5 was released in January of 2004, SMF 1.0 final was released in late December of 2004). If someone is still running YaBB SE at this point, it's not likely they're ever going to upgrade.

Sesquipedalian commented:
> OK, I snuck in a few more tests (yabb, 1.0 & 1.1), this is all I can do for today:
> 
> 3.0 Install OK 🍻 
> 3.0 => 3.0 FAIL 😥 
> 2.1 => 3.0 OK 🍻
> ...

For this PR we only care about 2.1 → 3.0 and 3.0 → 3.0, and specifically about the conversion to utf8mb4 for MySQL. There's no need to run tests on anything else at this point.

Regarding the 3.0 → 3.0, what went wrong?

sbulen commented:
> > OK, I snuck in a few more tests (yabb, 1.0 & 1.1), this is all I can do for today:
> > 3.0 Install OK 🍻
> > 3.0 => 3.0 FAIL 😥
> > 2.1 => 3.0 OK 🍻
> > ...
> 
> For this PR we only care about 2.1 → 3.0 and 3.0 → 3.0, and specifically about the conversion to utf8mb4 for MySQL. There's no need to run tests on anything else at this point.
> 
I would hope we plan on doing the same thing?  These upgrades used to work, & I think this is a clue the changes & new approach aren't working.  I suspect returning to the single command would work, rather than spreading it out over several steps.  Haven't tested that - I've been pretty busy lately with multiple RL challenges.

> Regarding the 3.0 → 3.0, what went wrong?
See the very first screenshot above for the browser result.  

CLI just returns & does nothing.

Sesquipedalian committed:
Move on to the right step after ConvertUtf8()

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Ensures default value for smf_log_errors.session column is correct

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian commented:
> > Regarding the 3.0 → 3.0, what went wrong?
>
> See the very first screenshot above for the browser result.
> 
> CLI just returns & does nothing.

Well, I can't reproduce that. 3.0 → 3.0 completes successfully for me, including the ConvertToUtf8() step, which is the only step we care about in this PR. I don't know what is causing 3.0 → 3.0 to fail for you right now, but since (a) it is happening at the first step and (b) you have run into the same problem [previously](https://github.com/SimpleMachines/SMF/pull/8459#issuecomment-2652508481), I don't think it is related to the changes in this PR.

> Uh... The 1.0 & 1.1 default for session is... uh... odd:

Although 1.0 and 1.1 issues are out of scope for this PR, I've added a fix for that in 9f3409d.

> 2.0 Latin1 => 3.0 FAIL 😥

Does it fail during ConvertToUtf8(), or somewhere else?


> Yabbse issue - didn't get very far at all... (Yabbse is so different, I wonder if it's time to punt on this support...):
>> - Updating Settings.php... FAILURE.

That'll be a problem to deal with in #8093

sbulen commented:
The 2.0 Latin 1 failure was highlighted up here - at first glance, it appears to be having issues with BLOBs?
https://github.com/SimpleMachines/SMF/pull/8425#issuecomment-2695441293

And yes, that's in the utf8mb4 conversion step.  I have not attempted a retest recently yet.

Note my 3.0 => 3.0 failure is due to $upcontext['language'] being blank on this line:
https://github.com/SimpleMachines/SMF/blob/783d8ebb5886e73b88b592bb587ce863335ae63a/other/upgrade.php#L1050

My steps:
 - I literally started from an empty directory & an empty DB
 - I installed 3.0, & did a quick test by making a post
 - I upgraded 3.0 => 3.0
 - PHP 8.4.2, MySQL 8.4.0, wamp

I haven't looked too closely, not a lot of time today...  My initial suspicion is there is a confusion between 'lang' and 'language', both are used at different points around there???  Or maybe Config language isn't set anywhere?

Note also that I don't think the throw error logic is working there, as the thrown error is not visible anywhere.  Just a 'try again' link via browser, or a clean exit via CLI.

Sesquipedalian commented:
> The 2.0 Latin 1 failure was highlighted up here - at first glance, it appears to be having issues with BLOBs? [#8425 (comment)](https://github.com/SimpleMachines/SMF/pull/8425#issuecomment-2695441293)
> 
> And yes, that's in the utf8mb4 conversion step. I have not attempted a retest recently yet.

Thanks, I guess I overlooked that. 🙂 

Unfortunately, I cannot reproduce that either. However, my test data for the 2.0 → 3.0 upgrade is just content generated by Populate.php, so perhaps that generated content isn't creating the right conditions to trigger the problem.

Could we arrange a way for me to get copies of the databases you are using, @sbulen? I would like to run tests using the same data as you are using so that I can try to figure out what the cause is.

In the meantime, though, I think I am going to go ahead and merge this for now. Even if there are still kinks to work out with the upgrader, waiting for the rest of the pending changes in this PR is holding up everything else in the development pipeline. Once I can get ahold of a copy of your test data and figure out the cause of the issues you are seeing, we will be able to fix the upgrade problems in the dedicated PR for upgrader changes.

sbulen commented:
OK.  I'll attempt to reproduce these two issues with current GH.  If I can reproduce, I'll write them up.

Note that for both issues, I started with a simple fresh install.  No other content.

Sesquipedalian commented:
> Note that for both issues, I started with a simple fresh install. No other content.

Hm. So, just to make sure that I understand correctly, the database you were using for the 2.0 → 3.0 upgrade was a fresh, empty install of 2.0 that you immediately upgraded to 3.0? 

If I am indeed understanding correctly, then the fact that I cannot reproduce the problem is _weird_.

sbulen commented:
> Hm. So, just to make sure that I understand correctly, the database you were using for the 2.0 → 3.0 upgrade was a fresh, empty install of 2.0 that you immediately upgraded to 3.0?

Yep.  latin1 db.
